### ISSUE TITLE
Provide a way to redirect stdout/stderr into a file

### ISSUE THREAD

**Issue by ash2k** [bazelbuild/bazel#5511]
### Description of the problem / feature request:

It is common for programs in unix world to output the result of an invocation to stdout. The current way to capture that output is to either use `ctx.actions.run_shell()` or a wrapping shell script to redirect it to a file. This requires spawning an extra process and hence adds some overhead. Perhaps it makes sense to provide a way for `ctx.actions.run()` to redirect stdout and/or stderr into a file.

### Feature requests: what underlying problem are you trying to solve with this feature?

I have a tool that does what I described above - writes the result of its invocation to stdout. I need to capture that output and put it into a file - it is the output of the rule. Currently I have to wrap it in a shell script and I thought it might be good to suggest this feature. Thanks.

**Comment by lfpino** [comment]
Assigning to @dslomov for triage.

**Comment by laurentlb** [comment]
I agree this would be useful. The current workaround is to use the shell. This is unfortunate because we need two implementations (bash / windows). So, something similar to https://github.com/bazelbuild/bazel-skylib/blob/master/rules/private/copy_file_private.bzl

You also need to be careful about quoting. We have https://github.com/bazelbuild/bazel-skylib/blob/master/lib/shell.bzl for the Linux shell, but nothing for cmd yet.

**Comment by laszlocsomor** [comment]
Quoting arguments is difficult with cmd. I typically put paths into envvars instead.

**Comment by joshua-leyshon-canva** [comment]
Has any progress been made on this? Or, can anyone share a general wrapper that could be used to achieve the same goal?

**Comment by alandonovan** [comment]
Is this feature necessary?  The `run_shell(cmd=...)` is an arbitrary shell program, and the shell is good at file redirection. To redirect the stdout of the script to one of the output files, then add this to the beginning of the shell command:

"""
exec > output
...
"""

Judging from the number of requests, this is clearly an obscure solution, but the problem is a lack of documentation, not features.

It's unfortunate that Bazel has any explicit dependency on Bash (or any Bourne shell), both because it is unnecessary (you can always exec\ `/bin/sh -c cmd`), and because it is not a standard tool on Windows. Ideally Bazel's run_shell command would be eliminated from the core (along with all other mentions of a shell---including genrule), and for those that really want shell scripting, not just a subprocess, an optional external library would provide a `bash_command` rule that adds a dependency on Bash-for-Windows.

**Comment by alandonovan** [comment]
I agree that extending ctx.run to allow redirecting the standard output to a file is a reasonable feature. I don't think it's necessary for stderr because its contents are unstructured, and by redirecting it, build failure messages would become hidden.

**Comment by aiuto** [comment]
Redirecting stderr is useful in the case where you are building tests for error messages. For example
- I have a tool
- I create a genrule that runs that tool with known bad input, and captures stderr to an output
- I add a diff_test or golden_file_test against that output.

**Comment by alandonovan** [comment]
> I create a genrule that runs that tool with known bad input, and captures stderr to an output

If you're creating a genrule, you're using a shell, so redirection is trivial.

**Comment by aiuto** [comment]
I'm jumping the gun a bit. What I would like to see out of genrule is the capability to raw run a command without having bash. Something more like an execv.

**Comment by alandonovan** [comment]
I agree, we should remove all dependencies on Bash from the core, which would mean replacing all uses of genrule with either:
(1) a function or rule that wraps process creation and sets envp, argv, and stdout, without a shell. This is possible today using ctx.actions.run, but we need to expose it to BUILD files in a form that doesn't require adding a new bzl file for each usage.
(2) a rule defined in a library that executes a Bash script, and works on Windows too (this means it would add a dependency). (Tony, I've sent you a quick sketch of # 2 in CL 320646234.)

**Comment by ElectricRCAircraftGuy** [comment]
It appears the output of bazel commands is already output to a file, no? In `command.log`, located some place like this: `/home/username/.cache/bazel/_bazel_username/3e8af127f8b488324cdf41111355ff4c/command.log`.

**Comment by alandonovan** [comment]
Alternative approaches:

1. "Process wrapper in skylib" (https://docs.google.com/document/d/13M9ArI_Qdj3PuJk2kXSRm6YISp8uceq_lEG8_I7kXcA)
    We should evaluate whether it's possible to implement this approach without the C++ toolchain dependency by writing the Windows implementation in the CMD.exe script language. (The code needn't be pleasant, just correct.)
2. "Windows: Bash-less alternative for genrule" (https://github.com/bazelbuild/bazel/issues/7503)
3. My Google internal pending change cr/320646234:

"""
bazel: sketch of bash_command, replacement for genrule

This CL contains a rough sketch of a design for a replacement of genrule
that meets the following criteria:

1) defined entirely in Starlark.
2) provides equivalent functionality to $(location) using Bash 4 hash tables.
3) allows the rule to add args and environment variables from Starlark values.
4) makes explicit the use of Bash. On Windows, this would entail a dependency
   to a ported executable.
...
"""

**Comment by cloudhan** [comment]
I am curious why a simple non-breaking, improve everyone's life feature, can be that hard to get addressed in every project. The ctx.actions side deside not to support stdout/err redirection. The bazel command insists on capture all action stdout and garbage fill users terminal with `INFO: From Executing blahblah`. And the [f**king vendor tool side insists on garbage fill users' stdout](https://developercommunity.visualstudio.com/t/allow-having-clexe-not-print-the-compiled-source-f/717761). ... 

If this feature is not implemented, once rules move to starlark, I'd like too see users terminal is always garbage filled, with `progress_message` buried in.

1. https://stackoverflow.com/questions/1453767/prevent-cl-exe-from-printing-the-compiled-source-file

**Comment by UlrichEckhardt** [comment]
Stumbled over the same issue. As alternative to (nonportable) shell code and (big & hairy) C++ compilers, there's also Python:
```python
from optparse import OptionParser
import os
import sys
import subprocess

if __name__ == '__main__':
    parser = OptionParser()
    parser.add_option('--output',
                      dest = 'output',
                      help = 'filename where to redirect stdout')
    parser.add_option('--verbose',
                      action = 'store_true',
                      default = False,
                      dest = 'verbose',
                      help = 'provide diagnostic output')
    (options, args) = parser.parse_args()

    # run actual program
    if options.verbose:
        print(f'Running binary (argv={repr(args)}).')
    process = subprocess.run(args,
                             capture_output = True,
                             universal_newlines = True)
    if options.verbose:
        print('Stdout:')
        print(''.join('  ' + l for l in process.stdout.splitlines(keepends = True)))
        print('Stderr:')
        print(''.join('  ' + l for l in process.stderr.splitlines(keepends = True)))
        print(f'Process finished with exit code {process.returncode}.')

    # write results to the output file
    with open(options.output, 'w') as f:
        f.write(process.stdout)

    sys.exit(process.returncode)
```
Put this wrapper into a `py_binary` and then you can use it in an action. Just use `[ctx.executable._the_wrapper, "--output", out_file.path, "--"]` as args, followed by the argv of the program you want to wrap. Note the `"--"`, which serves as a separator between the two argument lists.

**Comment by alanfalloon** [comment]
> Stumbled over the same issue. As alternative to (nonportable) shell code and (big & hairy) C++ compilers, there's also Python:

@UlrichEckhardt that is fine and effective Python, but I think in this case Python has a number of issues:
-  Far more systems ship with a Bourne compatible shell than a Python interpreter so carefully written shell code ([shellcheck](https://www.shellcheck.net) can help here) will be a more portable option.
- Startup times for Python are really bad, especially in Bazel. Even on a good day, Python takes a long time to even execute the first line of the module, but Bazel makes this far less efficient in BRE or sandboxing because it needs to drags in the Python interpreter and its massive set of stdlib files before it can even execute the action.

All this means that on short-running actions, your wrapper becomes the major cost. If one must use a wrapper, keeping it very fast to startup, with few runfiles, is key. A small shell script or preferably a statically linked native binary would be my choice.

However, all this is moot. The point of this issue is that it would be trivial for Bazel add input and output redirection to/from `File` objects in `ctx.actions.run` and that would be a huge usability improvement and be much more efficient than any wrapper.

**Comment by jonathan-enf** [comment]
Another advantage of the integral redirect would be that a build action that fails could then automatically dump the redirected log to the console, similar to the behavior of --test_output=errors.  This can all be accomplished by wrapper scripts, but getting it right efficiently and consistently should make this a first-class bazel feature.

**Comment by ilya-bobyr** [comment]
I think this is still a desired functionality.

**Comment by bcsgh** [comment]
I second the motion.

Just because the request is so fully described there's nothing more to say and the team hasn't gotten around to implementing it in 6 years doesn't make it stale.

**Comment by bcsgh** [comment]
_Still_ not stale. (And do you really want to set up rules that reward people for being impatient?)

**Comment by Wyverald** [comment]
> (And do you really want to set up rules that reward people for being impatient?)

If you're getting _really_ impatient, PRs are always welcome.

**Comment by adonovan** [comment]
> If you're getting really impatient, PRs are always welcome.

I think this issue still needs a clear consensus on the right implementation. At that point, a PR would indeed be welcome.

**Comment by bcsgh** [comment]
@adonovan I think there's already clear consensus on what people want to accomplish and I suspect almost nobody would take issue with _any_ reasonable implementation. From scanning the comments above, it looks like the only user relevant "open question" is if capturing stderr should be allowed which seems like it shouldn't be viewed as a blocker. (IMHO, if someone thinks they want to capture stderr, then nothing bazel does will prevent them from doing it, and most likely what they come up with will be worse for the user than what bazel could provide.)

**Comment by adonovan** [comment]
> @adonovan I think there's already clear consensus on what people want to accomplish and I suspect almost nobody would take issue with any reasonable implementation.

It seems like no-one disputes the addition of a `ctx.actions.run(stdout=File)` parameter.

Capturing stderr in the same way sounds reasonable to me, though of course stderr streams are unstructured and highly likely to be nondeterministic due to concurrency, log timestamps, etc, which would defeat action caching. If we add stderr, it should support using the same file as for stdout, using a single duplicated file descriptor.
 
Other open questions gleaned from the 50+ comments, at least according to Gemini 3, include:
```
   * Interaction with `execution_requirements`: How does redirection interact with
     remote execution? Some remote execution protocols might have specific ways of
     handling standard streams that Bazel needs to abstract over.
   * Buffering and Performance: There are questions about how Bazel should handle the
     streaming of this data to disk, especially for very large outputs, and whether it
     affects the "no-output-unless-error" philosophy of Bazel's UI.
   * API Signature Specifics:
       * Should there be a way to "tee" the output (both capture to a file and show on
         console)? Most consensus points towards only redirection to avoid complexity.
       * How to handle the case where a tool already writes to a file? (i.e., avoiding
         confusion between the tool's own flags and Bazel's redirection).
   * Implementation Complexity: A major open question is the plumbing required in
     Bazel's execution engine (the Spawn and Action layers) to support this across all
     spawn runners (local, sandboxed, remote).
```

My point is that this is a change to a key public interface that we will have to live with to years to come, and as such there should be a clearly stated consensus on the API change before we proceed to implementation.

**Comment by bcsgh** [comment]
Tl;dr; who should be consulted with regards to consensus? For the open question, I shouldn't be, and likely the same for most of the people watching this issue.

Those all sound like legitimate concerns. However, as a user of bazel (and just speaking for me) how they are resolved doesn't really matter to me; my opinion would at best be bike shedding.

I can't speak for anyone else, but I'd rather get a solution sooner even if that means the actual maintainers of bazel don't invite much feedback from the rules authors.

**Comment by adonovan** [comment]
> Tl;dr; who should be consulted with regards to consensus? For the open question, I shouldn't be, and likely the same for most of the people watching this issue.
> 
> Those all sound like legitimate concerns. However, as a user of bazel (and just speaking for me) how they are resolved doesn't really matter to me; my opinion would at best be bike shedding.
> 
> I can't speak for anyone else, but I'd rather get a solution sooner even if that means the actual maintainers of bazel don't invite much feedback from the rules authors.

Of course everyone wants things to move quickly in the direction they favor, but equally everyone is slowed down by all the little features that get added without regard to their cohesiveness. I assume the Starlark API folks on Google's Bazel team (@comius @brandjon @tetromino and others) will make the final decision, but that's really a question about Bazel governance.

Most programming languages have some kind of process--Java [JSRs](https://jcp.org/ja/jsr/overview), Python [PEPs](https://peps.python.org/), Go [proposals](https://go.googlesource.com/proposal/+/master/README.md)--for reviewing community-proposed interface changes independent (mostly) of implementation. All are imperfect, but they do help reduce casual blunders and incoherence (my favorite example being Bazel's three different syntaxes for expressing recursion over a tree). The Bazel project could benefit from something similar.

**Comment by bcsgh** [comment]
> Of course everyone wants things to move quickly in the direction they favor, but equally everyone is slowed down by all the little features that get added without regard to their cohesiveness.

My point is that "the direction I favor" is rather shallow and barely overlaps with the decisions that impact cohesiveness. I want a specific _ability_ and trust the owners enough that I expect any _implementation_ they ship will be good enough for me. Inviting more people to express opinions on those other details slows things down for little benefit.

**Comment by oakad** [comment]
How is this absolutely glaring omission is still open after all these years?
Wrapper scripts are evil and annoying, especially one liners for the sake of redirect.

**Comment by rdesgroppes** [comment]
If `bazel_lib` is an option for you, then it now provides a `run_binary(..., stdout=...)` for that purpose since bazel-contrib/bazel-lib#1259.

**Comment by oakad** [comment]
bazel_lib's run_binary pulls in an entire go toolchain (300MBs ?) to do a most mundane task which should be available out of the box.

Hardly an efficient approach - wrapper scripts don't look that bad in comparison. :-)

**Comment by fmeum** [comment]
Releases of `bazel-lib` ship with precompiled static binaries and do not depend on rules_go at all. 

Please keep in mind that a "native" solution to this would pretty much have to look the same. Since it needs to be able to run remotely while still supporting stdout/stderr redirection to the same file (which can't be emulated by capturing both separately), this can't be done via "Bazel in-process magic". I would expect every solution to ultimately come down to a wrapper script, it's just a question of who maintains that script and in which language it's written.

If you come up with a more efficient approach, please send a prototype as a PR and I'm happy to review and advocate for it.

**Comment by oakad** [comment]
I haven't seen any precompiled binaries in 3.3.1, it looked like it compiled the helpers from scratch.

The question of what constitutes a default toolchain for bazel is indeed an interesting one. I always thought it's Java with an assumption that whatever remote builders involved should have Java as well by default (this assumption is carried from other tools, such as Jenkins).

With this sort of thinking I imagined helper apps like these should be Java as well and ship bundled with Bazel.

**Comment by fmeum** [comment]
Java binaries are always a jar plus a full JDK, so that's arguably worse than a static Go binary. Bazel can't assume that a remote executor has anything installed, in particular not a full JDK.

**Comment by bcsgh** [comment]
Given that `[ctx.actions.run()`](https://bazel.build/rules/lib/builtins/actions.html#run) already needs some kind of worker that can "fork/exec" wherever stuff should be run the most robust solution would be for that worker to do the redirects. At a guess that worker is currently written in Java and it looks like it's rather easy to get Java to do redirection of `stdout`/`stderr` to the same or different files. The only downside to that is that it's dependent on the properties of the Bazel maintainers.

**Comment by fmeum** [comment]
This is true for local execution, but not for remote execution, both of which need to be supported for each new feature. The REAPI only provides stdout and stderr as separate digests (https://github.com/bazelbuild/remote-apis/blob/becdd8f9ff811df88a22d3eadd6341753d51d167/build/bazel/remote/execution/v2/remote_execution.proto#L1334), it doesn't provide a way to get both directed to the same output (preserving relative ordering). Please engage with the REAPI working group about a protocol extension if you want to see this.

**Comment by bcsgh** [comment]
So change the API. My point is that the only thing that makes this a hard problem is assuming that the core of Bazel won't be changed as part of the solution.

Also, 
1. Dumping both outputs to the same file is at most a nice to have, getting one or both as file(s) without a wrapper is the main ask.
2. If you can already get them back as an artifact (presumably for reporting errors) then allowing those to be files generated by an action seems like it should already be possible without things changing on the remote end.
3. Getting them as the same file could be done, under the existing result API, by the agent doing normal stream operations (triggered by a single flag in the request side of the API) and then either reporting the same artifact for both or reporting one as an empty file.

**Comment by fmeum** [comment]
> My point is that the only thing that makes this a hard problem is assuming that the core of Bazel won't be changed as part of the solution.

That's not a hard assumption at all. 

Since all process APIs in stdlibs I'm aware of support the "same file redirection" case naturally, I could see support for distinct stdout/stderr files only to be quite confusing.

If you think that a native solution is preferable to the `bazel-lib` wrapper approach the best path forward from my point of view would thus be to make a case for the small required REAPI protocol extension via an issue on the remote-apis repo and/or in person at the regular REAPI meeting (https://github.com/bazelbuild/remote-apis#api-community). If folks there are generally in favor of it, that would also make the argument for a Bazel API extension stronger.

**Comment by bcsgh** [comment]
So are you saying that the reason that this hasn't been getting any progress is that the proposal was made in the wrong component?

I think most people here (at least users of Bazel, if not maintainers) are under the impression that issues filed here are with respect to "Bazel and all it's bits" and that solutions that require changes to others parts of Bazel are automatically within scope without having to explicitly say that. (If that's not the case, then where is that assumption valid? Users shouldn't need to understand the inner workings of Bazel just to know where to file issues and FRs.)

Also: I agree that "redirect to same file" is highly desirable, but if requiring that prevents ever getting any redirection at all I'd toss it out a window in a heartbeat.

**Comment by fmeum** [comment]
Filing this FR here was entirely correct, it's just that making actual progress requires someone (Bazel team or external contributor) to actually spec out and get agreement on the feature (takes a short design doc) and to ensure that it's actually implementable (which would require some REAPI extension). The actual implementation is probably agentable at that point and if not, I'm happy to help fiddle with Bazel internals.

The REAPI extension could also be skipped if the feature is limited to just stdout for now, but the design would still need to discuss how stderr would fit if added in the future (the consistency that Alan talked about).

**Comment by bcsgh** [comment]
Tl;dr; I think the `stdout` only case could be added to the user API right now without much risk of creating future problems. It wouldn't fully resolve this FR, but I at least have no issue with releasing well designed features resolving only part of a FR.

From an end user API perspective I see basically only two options for the ideal final end state:

1. Add `stdout`/`stderr` arguments to `run()` that take File values as outputs with the special behavior that if they are the same object the streams are merged.
2. Like option 1 but where they must be unique files (or unset/`None`) and a third bool field (`combine_streams`?) triggers the merged case.

There might be an option 3 of just not supporting merged streams (which would likely make passing the same file twice an error).

I think that in all those cases, passing just `stdout` has identical semantics. That strongly implies that implementing just that case initially won't get in the way of whatever is chosen for the more complicated cases. Option 3 is also a subset of both 1 and 2 so a similar argument applies though I suspect it's weaker for that case and it also has some other considerations (e.g. it being a common source of non-deterministic output and thus more likely to be a foot gun than other cases, but that's not really a new thing).

The implementation end doesn't really matter *to me* so I won't comment on that.

Another bit: what about a `stdin`?

**Comment by fmeum** [comment]
@bcsgh I sent https://github.com/bazelbuild/bazel/pull/30408 to add the `stdout` parameter. I'm not planning to work on `stderr`/`stdin` as I believe those are better served by shell redirects (I am planning to improve the hermeticity of `run_shell`).
